guides/dsl.md

```diff
@@ -0,0 +1,241 @@
+# Eve JavaScript/TypeScript DSL
+
+The forthcoming v0.3 release of Eve supports a domain specific langauge (DSL) that interacts with the Eve server, allowing you to intermix Javascript and Eve code. There are several benefits to using the DSL:
+
+1) The syntax is native Javascript, so if you know JS you can write in the DSL
+2) You can use the DSL with Javascript functions and libraries
+3) The DSL can be used "a la carte", so you can use as much of Eve as you need to for your project. Thus you can write an entire application in Eve, just use it as a datastore, or anywhere inbetween. This makes it easy to integrate Eve with an existing Javascript application.
+4) You can easily import data into Eve. If you can get your data in Javascript, you can use it in Eve.
+
+## Using the DSL
+
+To use the DSL, import it into your application.
+
+```javascript
+import {Program} from "./runtime/dsl2";
+```
+
+Then you instantiate a new Eve program, naming it however you want.
+
+```javascript
+let program = new Program("program name");
+```
+
+Writing DSL code is similar to writing native Eve code; you attach blocks to the `program`, which search for records and then return records. Blocks in the DSL operate as a `bind` would in native Eve. Let's look at the DSL block syntax:
+
+```javascript
+  program.block("block description", ({find, record}) => {
+    // search for records tagged "person"
+    let person = find("person");
+    return [
+      // bind a record tagged "greeting" with an attribute "message"
+      record("greeting", {message: `Hello ${person.name}`})
+    ];
+  });
+```
+
+The equivalent Eve block would be
+
+```eve
+search
+  person = [#person]
+
+bind
+  [#greeting message: "Hello {{person.name}}"]
+```
+
+The identity of a `record()` is established by everything within the parentheses. Therefore, the record `record("greeting", {message: `Hello ${person.name}`})` will create one record for every `person.name`.
+
+If would like to commit records instead of binding them, you can use the `commit()` method in place of `block()`. The implicatino here is that you can no longer create blocks which both commit and bind records at the same time, which we've found is not a common case.
+
+The DSL supports methods which mirror a couple of the Eve update operators. They are:
+
+### add()
+
+ dds attribute/value pairs to a record. These added values do not contribute to the identity of a record, allowing you to add multiple values toa single record. This is equivalent to the Eve `+=` operator. For example:
```

Comment:
You can also use `.add()` on an entity variable, e.g. `friend.add("quality", "swell")`